#! /usr/bin/env python2

from dynamic_reconfigure.parameter_generator_catkin import *

PACKAGE = "usb_cam"

RECONFIGURE_CLOSE = 3 # Parameters that need a sensor to be stopped completely when changed
RECONFIGURE_STOP = 1 # Parameters that need a sensor to stop streaming when changed
RECONFIGURE_RUNNING = 0 # Parameters that can be changed while a sensor is streaming

gen = ParameterGenerator()

io_methods = gen.enum([gen.const("mmap",       str_t, "mmap", ""),
                       gen.const("read",       str_t, "read", ""),
                       gen.const("userptr",    str_t, "userptr", ""),
                       ], "IO method name")

pix_formats = gen.enum([gen.const("yuyv",      str_t, "yuyv", ""),
                        gen.const("uyvy",      str_t, "uyvy", ""),
                        gen.const("mjpeg",     str_t, "mjpeg", ""),
                        gen.const("yuvmono10", str_t, "yuvmono10", ""),
                        gen.const("rgb24",     str_t, "rgb24", ""),
                        ], "Pixel format")

#       Name,                 Type,     Reconfig level,      Description, Default, Min, Max
gen.add("brightness",         int_t,    RECONFIGURE_RUNNING, "Brightness", -1, -1, 255)
gen.add("contrast",           int_t,    RECONFIGURE_RUNNING, "Contrast", -1, -1, 255)
gen.add("saturation",         int_t,    RECONFIGURE_RUNNING, "Saturation", -1, -1, 255)
gen.add("sharpness",          int_t,    RECONFIGURE_RUNNING, "Sharpness", -1, -1, 255)
gen.add("io_method_name",     str_t,    RECONFIGURE_CLOSE,   "IO method", "mmap", edit_method=io_methods)
gen.add("image_width",        int_t,    RECONFIGURE_RUNNING, "Image width", 640)
gen.add("image_height",       int_t,    RECONFIGURE_RUNNING, "Image height", 480)
gen.add("framerate",          int_t,    RECONFIGURE_RUNNING, "Framerate", 30)
gen.add("pixel_format_name",  str_t,    RECONFIGURE_RUNNING, "Pixel format", "mjpeg", edit_method=pix_formats)
gen.add("autofocus",          bool_t,   RECONFIGURE_RUNNING, "Autofocus", False)
gen.add("focus",              int_t,    RECONFIGURE_RUNNING, "Focus", -1)
gen.add("autoexposure",       bool_t,   RECONFIGURE_RUNNING, "Autoexposure", True)
gen.add("exposure",           int_t,    RECONFIGURE_RUNNING, "Exposure", 100)
gen.add("gain",               int_t,    RECONFIGURE_RUNNING, "Gain", -1, -1, 100)
gen.add("auto_white_balance", bool_t,   RECONFIGURE_RUNNING, "Auto white balance", True)
gen.add("white_balance",      int_t,    RECONFIGURE_RUNNING, "White balance", 4000)
gen.add("camera_frame_id",    str_t,    RECONFIGURE_RUNNING, "Frame ID", "head_camera")
gen.add("camera_name",        str_t,    RECONFIGURE_RUNNING, "Name", "")
gen.add("camera_info_url",    str_t,    RECONFIGURE_RUNNING, "Info URL", "")

exit(gen.generate(PACKAGE, "usb_cam_node", "V4LParameters"))

